util/limiter/limiter.go

```diff
@@ -0,0 +1,146 @@
+// unbounded memory growth, Limiter actually only tracks limits
+// precisely for the N most recently seen keys, and assumes that
+// untracked keys are well-behaved. This trades off absolute precision
+// for bounded memory use, while still enforcing well for outlier
+// keys.
+//
+// As such, Limiter should only be used in situations where "rough"
+// enforcement of outliers only is sufficient, such as throttling
+// egregious outlier keys (e.g. something sending 100 queries per
+// second, where everyone else is sending at most 5).
+//
+// Each key's token bucket behaves like a regular token bucket, with
+// the added feature that a bucket's token count can optionally go
+// negative. This implements a form of "cooldown" for keys that exceed
+// the rate limit: once a key starts getting denied, it must stop
+// requesting tokens long enough for the bucket to return to a
+// positive balance. If the key keeps hammering the limiter in excess
+// of the rate limit, the token count will remain negative, and the
+// key will not be allowed to proceed at all. This is in contrast to
+// the classic token bucket, where a key trying to use more than the
+// rate limit will get capped at the limit, but can still occasionally
+// consume a token as one becomes available.
+type Limiter[K comparable] struct {
+	// Size is the number of keys to track. Only the Size most
+	// recently seen keys have their limits enforced precisely, older
+	// keys are assumed to not be querying frequently enough to bother
+	// tracking.
+	Size int
+
+	// Max is the number of tokens available for a key to consume
+	// before time-based rate limiting kicks in. An unused limiter
+	// regains available tokens over time, up to Max tokens. A newly
+	// tracked key initially receives Max tokens.
+	Max int64
+	// Overdraft is the amount of additional tokens a key can be
+	// charged for when it exceeds its rate limit. Each additional
+	// request issued for the key charges one unit of overdraft, up to
+	// this limit. Overdraft tokens are refilled at the normal rate,
+	// and must be fully repaid before any tokens become available for
+	// requests.
+	//
+	// A non-zero Overdraft results in "cooldown" behavior: with a
+	// normal token bucket that bottoms out at zero tokens, an abusive
+	// key can still consume one token every RefillInterval
+	// interval. With a non-zero overdraft, a throttled key must stop
+	// requesting tokens entirely for a cooldown period, otherwise
+	// they remain perpetually in debt and cannot proceed at all.
+	Overdraft int64
+	// RefillInterval is the interval at which a key regains tokens for
+	// use, up to Max tokens.
+	RefillInterval time.Duration
+
+	mu    sync.Mutex
+	cache *lru.Cache[K, *bucket]
+}
+
+type bucket struct {
+	cur        int64     // current available tokens
+	lastUpdate time.Time // last timestamp at which cur was updated
+}
+
+// Allow reports whether the key can perform an action.
```

Comment:
Perhaps also: "…and if so charges the key one token."

Reply:
That's not correct though, in the presence of a non-zero Overdraft a denial also consumes a token (albeit an overdraft one). Given the entire purpose of the limiter is to do the accounting on whether or not actions are allowed, "... and also does bookkeeping" seems like the default behavior that isn't worth mentioning? If an exported API allowed querying _without_ doing bookkeeping, IMO that would be noteworthy, but not the reverse? Dunno, thoughts?

Reply:
Good point, re: correctness. Mostly what I was trying to get at was the idea that calling this also charges against the rate. So perhaps instead:

```go
// Allow charges the key one token (up to the overdraft limit) and reports whether
// that key can perform an action.
```

?

Reply:
sg, done.